### PR TITLE
Allow to set the icon position on Balloon

### PR DESCRIPTION
**Pull request by andrespaez90**
## New feature
  - Allow to set the icon position on Balloon: Now you can specify the position of the drawable (left, top, right or bottom)
  - Improve layout_ballon: remove LinearLayout and add a custom textView

### CODE REVIEW

File: balloon/src/main/java/com/skydoves/balloon/IconForm.kt [hunk 1]
```diff
@@ -37,8 +37,9 @@ inline fun iconForm(context: Context, block: IconForm.Builder.() -> Unit): IconF
  * for customizing popup icons easily.
  */
 class IconForm(builder: Builder) {
-
   val drawable = builder.drawable
+  var drawableRes = builder.drawableRes
```
Comment (skydoves) [inline balloon/src/main/java/com/skydoves/balloon/IconForm.kt:41]: Seems to missed `@DrawableRes` annotation.

File: balloon/src/main/java/com/skydoves/balloon/IconForm.kt [hunk 2]
```diff
@@ -48,6 +49,10 @@ class IconForm(builder: Builder) {
   class Builder(val context: Context) {
     @JvmField
     var drawable: Drawable? = null
+    @JvmField
```
Comment (skydoves) [inline balloon/src/main/java/com/skydoves/balloon/IconForm.kt:52]: Seems to missed @DrawableRes annotation. (2)

File: balloon/src/main/java/com/skydoves/balloon/IconForm.kt [hunk 3]
```diff
@@ -60,9 +65,13 @@ class IconForm(builder: Builder) {
 
     /** sets the [Drawable] of the icon using resource. */
     fun setDrawableResource(@DrawableRes value: Int): Builder = apply {
-      this.drawable = context.contextDrawable(value)
+      this.drawableRes = value
     }
 
+    /** sets gravity ofthe [Drawable] of the icon using resource. */
```
Comment (skydoves) [inline balloon/src/main/java/com/skydoves/balloon/IconForm.kt:71]:
```suggestion
    /** sets gravity of the [Drawable] of the icon using resource. */
```
A tiny typo.

File: balloon/src/main/java/com/skydoves/balloon/TextViewExtension.kt [hunk 4]
```diff
@@ -41,3 +46,41 @@ private fun fromHtml(text: String): Spanned? {
     Html.fromHtml(text)
   }
 }
+
+fun TextView.applyDrawable(vectorTextViewModel: VectorTextViewModel) {
```
Comment (skydoves) [inline balloon/src/main/java/com/skydoves/balloon/TextViewExtension.kt:50]: Seems to missed `internal` keyword.

File: balloon/src/main/java/com/skydoves/balloon/TextViewExtension.kt [hunk 5]
```diff
@@ -41,3 +46,41 @@ private fun fromHtml(text: String): Spanned? {
...
+    ?: vectorTextViewModel.heightRes?.let { context.resources.getDimensionPixelSize(it) }
+    ?: vectorTextViewModel.squareSizeRes?.let { context.resources.getDimensionPixelSize(it) }
+
+  val width = vectorTextViewModel.iconSize
+    ?: vectorTextViewModel.widthRes?.let { context.resources.getDimensionPixelSize(it) }
+    ?: vectorTextViewModel.squareSizeRes?.let { context.resources.getDimensionPixelSize(it) }
+
+  val drawableLeft: Drawable? =
+    vectorTextViewModel.drawableLeft ?: vectorTextViewModel.drawableLeftRes
+      ?.let {
```
Comment (skydoves) [inline balloon/src/main/java/com/skydoves/balloon/TextViewExtension.kt:61]:
```suggestion
 vectorTextViewModel.drawableLeft ?: vectorTextViewModel.drawableLeftRes?.let {
```
How about this? This is more explicit.

File: balloon/src/main/java/com/skydoves/balloon/custom/VectorTextViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package com.skydoves.balloon.custom
+
+import android.graphics.drawable.Drawable
+import androidx.annotation.ColorRes
+import androidx.annotation.DimenRes
+import androidx.annotation.Px
+
+data class VectorTextViewModel(
```
Comment (skydoves) [inline balloon/src/main/java/com/skydoves/balloon/custom/VectorTextViewModel.kt:24]: I think this name will make some confusion later.
If we read this code later without knowing about the `VectorTextView`, the name could be interpreted kind of `ViewModel`. How about `VectorTextViewParams`?

File: balloon/src/main/java/com/skydoves/balloon/extensions/DrawableExtension.kt [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import android.graphics.Canvas
+import android.graphics.drawable.BitmapDrawable
+import android.graphics.drawable.Drawable
+import androidx.annotation.ColorInt
+import androidx.annotation.ColorRes
+import androidx.annotation.Px
+import androidx.core.content.ContextCompat
+import androidx.core.graphics.drawable.DrawableCompat
+
+fun Drawable.tint(context: Context, @ColorRes tintColorRes: Int?): Drawable =
```
Comment (skydoves) [inline balloon/src/main/java/com/skydoves/balloon/extensions/DrawableExtension.kt:31]: Missing `internal` keyword.

File: balloon/src/main/java/com/skydoves/balloon/extensions/DrawableExtension.kt [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  tintColorRes?.let { tintWithColor(ContextCompat.getColor(context, it)) } ?: this
+
+fun Drawable.tintWithColor(@ColorInt tintColor: Int): Drawable = apply {
```
Comment (skydoves) [inline balloon/src/main/java/com/skydoves/balloon/extensions/DrawableExtension.kt:34]: Missing `internal` keyword. (2)

File: balloon/src/main/java/com/skydoves/balloon/extensions/DrawableExtension.kt [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  DrawableCompat.wrap(this)
+    .apply { DrawableCompat.setTintList(this, ColorStateList.valueOf(tintColor)) }
+}
+
+fun Drawable.resize(context: Context, @Px width: Int?, @Px height: Int?): Drawable =
```
Comment (skydoves) [inline balloon/src/main/java/com/skydoves/balloon/extensions/DrawableExtension.kt:39]: Missing `internal` keyword. (3)

**Comment by Tiam-Abderezai** [comment]
@skydoves  is it possible to pass Top AND End to setIconGravity()? I can either do Top, Bottom, Start, or End, not a mix of Top|End

**Pull request merged**